src/Stratis.Bitcoin/Base/ConsensusManagerBehavior.cs

```diff
@@ -0,0 +1,434 @@
+
+            // Ignoring "getheaders" from peers because node is in initial block download unless the peer is whitelisted.
+            if (this.initialBlockDownloadState.IsInitialBlockDownload() && !peer.Behavior<ConnectionManagerBehavior>().Whitelisted)
+            {
+                this.logger.LogTrace("(-)[IGNORE_ON_IBD]");
+                return;
+            }
+
+            var headers = new HeadersPayload();
+            ChainedHeader consensusTip = this.consensusManager.Tip;
+
+            ChainedHeader fork = this.chain.FindFork(getHeadersPayload.BlockLocators);
+            if (fork != null)
+            {
+                if ((consensusTip == null) || (fork.Height > consensusTip.Height))
+                {
+                    // Fork not yet validated.
+                    fork = null;
+                }
+
+                if (fork != null)
+                {
+                    foreach (ChainedHeader header in this.chain.EnumerateToTip(fork).Skip(1))
+                    {
+                        if (header.Height > consensusTip.Height)
+                            break;
+
+                        headers.Headers.Add(header.Header);
+                        if ((header.HashBlock == getHeadersPayload.HashStop) || (headers.Headers.Count == 2000))
+                            break;
+                    }
+                }
+            }
+
+            await peer.SendMessageAsync(headers).ConfigureAwait(false);
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// Processes "headers" message received from the peer.
+        /// </summary>
+        /// <param name="peer">Peer from which the message was received.</param>
+        /// <param name="headersPayload">Payload of "headers" message to process.</param>
+        /// <remarks>
+        /// "headers" message is sent in response to "getheaders" message or it is solicited
+        /// by the peer when a new block is validated (unless in IBD).
+        /// <para>
+        /// When we receive "headers" message from the peer, we can adjust our knowledge
+        /// of the peer's view of the chain. We update its pending tip, which represents
+        /// the tip of the best chain we think the peer has.
+        /// </para>
+        /// <para>
+        /// If we receive a valid header from peer which work is higher than the work
+        /// of our best chain's tip, we update our view of the best chain to that tip.
+        /// </para>
+        /// </remarks>
+        private async Task ProcessHeadersAsync(INetworkPeer peer, HeadersPayload headersPayload)
+        {
+            this.logger.LogTrace("({0}:'{1}',{2}:'{3}')", nameof(peer), peer.RemoteSocketEndpoint, nameof(headersPayload), headersPayload);
+
+            List<BlockHeader> headers = headersPayload.Headers;
+
+            if (headers.Count == 0)
```

Comment:
`!headers.Any()`

Reply:
LINQ is bad 🙂 
I'm trying to avoid it as much as possible. 

Here it will do more work that it's supposed to so `.Count == 0` is cheaper

Reply:
>LINQ is bad 
No LINQ is not bad generally, but I understand that in consensus code we want to avoid LINQ as a rule.

Reply:
I'll rephrase: LINQ is not bad, it's a gun with which you can shoot yourself in the leg. If you use it appropriately- it's good. If you are using it everywhere where it can be used- it's highly less efficient in comparison to if you'd write it yourself without linq.

Reply:
Here with Any it's not the worst case, but for example: 

https://referencesource.microsoft.com/#system.core/system/linq/Enumerable.cs

```
public static bool Any<TSource>(this IEnumerable<TSource> source) {
    if (source == null) throw Error.ArgumentNull("source");
    using (IEnumerator<TSource> e = source.GetEnumerator()) {
        if (e.MoveNext()) return true;
    }
    return false;
}
```

This guy will check if source is null. But I don't need it- I already know it's not null. Also it will create a disposable object, do smth with it and dispose it later. Kinda too much comparing to checking if count is 0.

Reply:
> !headers.Any()

no - https://stackoverflow.com/questions/5741617/listt-any-or-count 
https://stackoverflow.com/questions/305092/which-method-performs-better-any-vs-count-0

+ LINQ is bad